checks/frozen_deps.go

```diff
@@ -46,14 +46,102 @@ func FrozenDeps(c *checker.CheckRequest) checker.CheckResult {
 		isPackageManagerLockFilePresent,
 		isGitHubActionsWorkflowPinned,
 		isDockerfilePinned,
+		isDockerfileFreeOfInsecureDownloads,
 	)(c)
 }
 
-// TODO(laurent): need to support GCB
+// TODO(laurent): need to support GCB pinning.
+
+func isDockerfileFreeOfInsecureDownloads(c *checker.CheckRequest) checker.CheckResult {
+	return CheckFilesContent(CheckFrozenDeps, "*Dockerfile*", false, c, validateDockerfileDownloads)
+}
+
+func validateDockerfileDownloads(pathfn string, content []byte,
+	logf func(s string, f ...interface{})) (bool, error) {
+	contentReader := strings.NewReader(string(content))
+	res, err := parser.Parse(contentReader)
+	if err != nil {
+		return false, fmt.Errorf("cannot read dockerfile content: %w", err)
+	}
+
+	ret := true
+	files := make(map[string]bool)
+
+	// Walk the Dockerfile's AST.
+	for _, child := range res.AST.Children {
+		cmdType := child.Value
+		// Only look for the 'RUN' command.
+		if cmdType != "run" {
+			continue
+		}
+
+		var valueList []string
+		for n := child.Next; n != nil; n = n.Next {
+			valueList = append(valueList, n.Value)
+		}
+
+		if len(valueList) != 1 {
+			return false, ErrParsingDockerfile
+		}
+
+		// Validate it's not downloading and piping into a shell, like
+		// `curl | bash` (supports `sudo`).
+		r, err := validateCommandIsNotFetchPipeExecute(valueList[0], pathfn, logf)
+		if err != nil {
+			return false, err
+		} else if !r {
+			ret = false
+		}
+
+		// Validate it is not a download command followed by
+		// an execute: `curl > /tmp/file && /tmp/file`
+		//			   `curl > /tmp/file && bash /tmp/file`
+		//			   `curl > /tmp/file; bash /tmp/file`
+		//			   `curl > /tmp/file; /tmp/file`
+		// (supports `sudo`).
+		r, err = validateCommandIsNotFetchToFileExecute(valueList[0], pathfn, logf)
+		if err != nil {
+			return false, err
+		} else if !r {
+			ret = false
+		}
+
+		// Validate it's not shelling out by redirecting input to stdin, like
+		// `bash <(wget -qO- http://website.com/my-script.sh)`. (supports `sudo`).
+		r, err = validateCommandIsNotFetchToStdinExecute(valueList[0], pathfn, logf)
+		if err != nil {
+			return false, err
+		} else if !r {
+			ret = false
+		}
+
+		// TODO(laurent): add check for cat file | bash
+		// TODO(laurent): detect downloads of zip/tar files containing scripts.
+		// TODO(laurent): detect command being an env variable
```

Comment:
I think we should just try to capture the most common anti patterns, rather than trying to detect every single possible case, which doesn't seem too feasible or maintainable.

Reply:
agreed. I've added them for future enhancements.